encoding/codecv4.go

```diff
@@ -0,0 +1,624 @@
+package encoding
+
+import (
+	"crypto/sha256"
+	"encoding/binary"
+	"encoding/hex"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"math"
+	"math/big"
+	"reflect"
+	"sync/atomic"
+
+	"github.com/scroll-tech/go-ethereum/common"
+	"github.com/scroll-tech/go-ethereum/core/types"
+	"github.com/scroll-tech/go-ethereum/crypto"
+	"github.com/scroll-tech/go-ethereum/crypto/kzg4844"
+	"github.com/scroll-tech/go-ethereum/log"
+
+	"github.com/scroll-tech/da-codec/encoding/zstd"
+)
+
+type DACodecV4 struct {
+	enableCompress uint32
+}
+
+// Codecv4MaxNumChunks is the maximum number of chunks that a batch can contain.
+const Codecv4MaxNumChunks = 45
+
+// Version returns the codec version.
+func (o *DACodecV4) Version() CodecVersion {
+	return CodecV4
+}
+
+// NewDABlock creates a new DABlock from the given Block and the total number of L1 messages popped before.
+func (o *DACodecV4) NewDABlock(block *Block, totalL1MessagePoppedBefore uint64) (DABlock, error) {
```

Comment:
To reduce some duplicated code, how about embedding `DACodecV3` and overriding the functions that change or having a private instance of `DACodecV3` and forwarding the calls to functions that are exactly the same?

For this to work `DACodecV3` needs to embed `DACodecV2`, `DACodecV2` -> `DACodecV1` and `DACodecV1` -> `DACodecV0`.

Reply:
embed logic is added in [8dbf796](https://github.com/scroll-tech/da-codec/pull/25/commits/8dbf796031085af8a34921b8b3ab156c7a3427fb).